openwisp_controller/config/api/filters.py

```diff
@@ -0,0 +1,145 @@
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+DeviceGroup = load_model('config', 'DeviceGroup')
+
+
+class BaseConfigAPIFilter(OrganizationManagedFilter):
+    def _set_valid_filterform_lables(self):
+        # When not filtering on a model field, an error message
+        # with the label "[invalid_name]" will be displayed in filter form.
+        # To avoid this error, we need to provide the label explicitly.
+        raise NotImplementedError
+
+
+class TemplateListFilter(BaseConfigAPIFilter):
+    created__gte = filters.DateTimeFilter(
+        field_name='created',
+        lookup_expr='gte',
+    )
+    created__lt = filters.DateTimeFilter(
+        field_name='created',
+        lookup_expr='lt',
+    )
+
+    def _set_valid_filterform_lables(self):
+        self.filters['backend'].label = _('Template backend')
+        self.filters['type'].label = _('Template type')
+
+    def __init__(self, *args, **kwargs):
+        super(TemplateListFilter, self).__init__(*args, **kwargs)
+        self._set_valid_filterform_lables()
+
+    class Meta(BaseConfigAPIFilter.Meta):
+        model = Template
+        fields = BaseConfigAPIFilter.Meta.fields + [
+            'backend',
+            'type',
+            'default',
+            'required',
+            'created',
+        ]
+
+
+class VPNListFilter(BaseConfigAPIFilter):
+    def _set_valid_filterform_lables(self):
+        self.filters['backend'].label = _('VPN Backend')
+        self.filters['subnet'].label = _('VPN Subnet')
+
+    def __init__(self, *args, **kwargs):
+        super(VPNListFilter, self).__init__(*args, **kwargs)
+        self._set_valid_filterform_lables()
+
+    class Meta(BaseConfigAPIFilter.Meta):
+        model = Vpn
+        fields = BaseConfigAPIFilter.Meta.fields + ['backend', 'subnet']
+
+
+class DeviceListFilterBackend(DjangoFilterBackend):
+    def filter_queryset(self, request, queryset, view):
+        """
+        Validate that the request parameters contain
+        a valid configuration template uuid format
+        """
+        config_template_uuid = request.query_params.get('config__templates')
+        if config_template_uuid:
+            try:
+                # Attempt to convert the uuid string to a UUID object
+                config_template_uuid_obj = UUID(config_template_uuid)
+            except ValueError:
+                raise ValidationError({'config__templates': _('Invalid UUID format')})
+            # Add the config__templates filter to the queryset
+            return queryset.filter(config__templates=config_template_uuid_obj)
+        return super().filter_queryset(request, queryset, view)
+
+
+class DeviceListFilter(BaseConfigAPIFilter):
+    # Using filter query param name `with_geo`
+    # which is similar to admin filter
+    with_geo = filters.BooleanFilter(
+        field_name='devicelocation', method='filter_devicelocation'
+    )
```

Comment:
There is a problem here. The `config` app does not know about `geo`.
Look at how the admin filter is implemented, its code is in `openwisp_controller.geo`, not in `openwisp_controller.config`:

https://github.com/openwisp/openwisp-controller/blob/2b978723bf793aa40c88bd2d8f8e9bc8cd26c1bf/openwisp_controller/geo/admin.py#L85-L103

Adding this code here creates issue when `geo` is not installed.

Reply:
@nemesisdesign I attempted to add DeviceLocationFilter with similar logic to what was used in `admin.py`, but it did not work. Therefore, I used a different approach in commit [`2abc007` (#748)](https://github.com/openwisp/openwisp-controller/pull/748/commits/2abc0071169317b21e5bd72ecf3d3b36f6d7b654), and it is now working as expected. Please see the details below:

#### When `geo` is not installed (device location filter is absent)

#### DRF browsable API

![Screenshot from 2023-04-24 15-40-17](https://user-images.githubusercontent.com/56113566/233969176-61c94f41-cd6c-428c-9bc4-8e4f26de01a9.png)

#### Admin
![Screenshot from 2023-04-24 15-40-29](https://user-images.githubusercontent.com/56113566/233969185-e3a1886a-314c-48ab-a64b-e153fdbfdb0e.png)